File: server/node/test/hz/message/message_test.js
#!/usr/bin/env node

'use strict';

const assert = require('assert');
const message = require('../../../src/hz/message/message.js');

describe('message.Message', function() {
	beforeEach(function() {
        const hdr = new TestHolder();
        this._pst = new TestPost();
		this._msg = new message.Message(hdr, this._pst);
	});

	afterEach(function() {
		this._msg.close(function(err) {
            assert(!err);
        });
	});
  
	it('message.Message.request()', function() {
		this.timeout(0);
		
		const _this = this;

		// 测试{type: 'getpost_count'}
        _this._msg.request({type: 'get_post_count'}, function(err, res) {
            assert(!err);
			_this._pst.getCount(function callback(gerr, gres) {
                assert(!gerr);
				assert.deepEqual(gres, res);
			});
		});

		// 测试{type: 'get_allpost'}
        _this._msg.request({type: 'get_all_post'}, function(err, res) {
            assert(!err);
			_this._pst.getAll(function callback(gerr, gres) {
                assert(!gerr);
				assert.deepEqual(gres, res);
			});
		});

		// 测试{type: 'get_post_by_id'}
        _this._msg.request({type: 'get_post_by_id'}, function(err, res) {
            assert(!err);
			_this._pst.getById({}, function callback(gerr, gres) {
                assert(!gerr);
				assert.deepEqual(gres, res);
			});
		});

		// 测试{type: 'add_post'}
        _this._msg.request({type: 'add_post'}, function(err, res) {
            assert(!err);
			_this._pst.add({}, function callback(aerr, ares) {
                assert(!aerr);
				assert.deepEqual(ares, res);
			});
		});

		// 测试{type: 'modify_post'}
        _this._msg.request({type: 'modify_post'}, function(err, res) {
            assert(!err);
			_this._pst.modify({}, function callback(merr, mres) {
                assert(!merr);
				assert.deepEqual(mres, res);
			});
		});

		// 测试{type: 'remove_post'}
        _this._msg.request({type: 'remove_post'}, function(err, res) {
            assert(!err);
			_this._pst.remove({}, function callback(rerr, rres) {
                assert(!rerr);
				assert.deepEqual(rres, res);
			});
		});

		// 测试{type: 'other'}
        _this._msg.request({type: 'other'}, function(err, res) {
            assert(!err);
			assert.equal(res.status, 'invalid_parameter');
		});
	});
});

/**
 * 模拟holder类
 */
function TestHolder() {
}

TestHolder.prototype.inst = function() {
}

TestHolder.prototype.close = function(callback) {
	callback(undefined);
}

TestHolder.prototype.destroy = function(callback) {
	callback(undefined);
}

/**
 * 模拟post类
 */
function TestPost() {
}

TestPost.prototype.getCount = function(callback) {
	callback(undefined, {status:'ok', data:{'count':1234}});
}

TestPost.prototype.getAll = function(callback) {
	callback(undefined, {status:'ok', data:[{id:1234, timestamp:123456, content:'abcd'}, {id:12345, timestamp:1234567, content:'abcde'}]});
}

TestPost.prototype.getById = function(data, callback) {
	callback(undefined, {status:'ok', data:{id:1234, timestamp:123456, content:'abcd'}});
}

TestPost.prototype.add = function(data, callback) {
	callback(undefined, {status:'ok', data:{id:1234, timestamp:123456, content:'abcd'}});
}

TestPost.prototype.modify = function(data, callback) {
	callback(undefined, {status:'ok', data:{id:1234, timestamp:123456, content:'abcd'}});
}

TestPost.prototype.remove = function(data, callback) {
	callback(undefined, {status:'ok', data:{id:1234}});
}

File: server/java/test/hz/message/MessageTest.java
package hz.message;

import static org.junit.Assert.*;

import org.junit.After;
import org.junit.Before;
import org.junit.Test;
import org.json.JSONObject;

/**
 * 测试Message
 */
public class MessageTest {
	/** message对象 */
	private Message _message;
	/** post对象 */
	private Post _post;
	
	@Before
	public void before() {
        Holder holder = new TestHolder();
        _post = new TestPost();
        _message = new Message(holder, _post);
	}
	
	@After
	public void after() throws Exception {
		_message.close();
	}

	/**
	 * 测试Message.request()
	 * 
	 * @throws Exception 错误异常
	 */
	@Test
    public void test_request() throws Exception {
        // 测试{"type": "get_post_count"}
        assertEquals(_message.request(new JSONObject("{\"type\": \"get_post_count\"}")).toMap(), _post.getCount().toMap());
        
        // 测试{"type": "get_all_post"}
        assertEquals(_message.request(new JSONObject("{\"type\": \"get_all_post\"}")).toMap(), _post.getAll().toMap());
        
        // 测试{"type": "get_post_by_id"}
        assertEquals(_message.request(new JSONObject("{\"type\": \"get_post_by_id\"}")).toMap(), _post.getById(new JSONObject("{}")).toMap());
        
        // 测试{"type": "add_post"}
        assertEquals(_message.request(new JSONObject("{\"type\": \"add_post\"}")).toMap(), _post.add(new JSONObject("{}")).toMap());
        
        // 测试{"type": "modify_post"}
        assertEquals(_message.request(new JSONObject("{\"type\": \"modify_post\"}")).toMap(), _post.modify(new JSONObject("{}")).toMap());
        
        // 测试{"type": "remove_post"}
        assertEquals(_message.request(new JSONObject("{\"type\": \"remove_post\"}")).toMap(), _post.remove(new JSONObject("{}")).toMap());
        
        // 测试{"type": "other"}
        assertEquals(_message.request(new JSONObject("{\"type\": \"other\"}")).getString("status"), "invalid_parameter");
	}

	/**
	 * 模拟holder类
	 */
	private static class TestHolder implements Holder {
		@Override
	    public Object inst() {
	        return null;
		}
	
		@Override
	    public void destroy() throws Exception {
		}
	
		@Override
	    public void close() throws Exception {
		}
	}

	/**
	 * 模拟post类
	 */
	private static class TestPost implements Post {
		@Override
	    public JSONObject getCount() throws Exception {
	        return new JSONObject("{\"status\": \"ok\", \"data\": {\"count\": 1234}}");
		}
	
		@Override
	    public JSONObject getAll() throws Exception {
	        return new JSONObject("{\"status\": \"ok\", \"data\": [{\"id\": 1234, \"timestamp\": 123456, \"content\": \"abcd\"}, {\"id\": 12345, \"timestamp\": 1234567, \"content\": \"abcde\"}]}");
		}
	
		@Override
	    public JSONObject getById(JSONObject data) throws Exception {
	        return new JSONObject("{\"status\": \"ok\", \"data\": {\"id\": 1234, \"timestamp\": 123456, \"content\": \"abcd\"}}");
		}
	
		@Override
	    public JSONObject add(JSONObject data) throws Exception {
	        return new JSONObject("{\"status\": \"ok\", \"data\": {\"id\": 1234, \"timestamp\": 123456, \"content\": \"abcd\"}}");
		}
	
		@Override
	    public JSONObject modify(JSONObject data) throws Exception {
	        return new JSONObject("{\"status\": \"ok\", \"data\": {\"id\": 1234, \"timestamp\": 123456, \"content\": \"abcd\"}}");
		}
	
		@Override
	    public JSONObject remove(JSONObject data) throws Exception {
	        return new JSONObject("{\"status\": \"ok\", \"data\": {\"id\": 1234}}");
		}
	}
}
